### ISSUE TITLE
Transform shifting clusters on repeated calls

### ISSUE THREAD

**Issue by Kenneth-Schroeder** [lmcinnes/umap#247]
Hi @lmcinnes,
I am trying to use UMAP to show the increasing ability to seperate different classes over the training process of a neural network, similar to [https://medium.com/@rakesh.chada/understanding-neural-networks-by-embedding-hidden-representations-f256842ebf3a](url)

My idea was to select an epoch e, a layer l, as well as an input i.
I would then fit a UMAP instance on the high dimensional representations of i at layer l for that epoch e.
Afterwards, I want to use the learned transformation on other epochs for a fixed i and l.

However, the transformations for the same e, l and i seem to change after several runs on other e's. I don't mean some noisy variations, that I would expect to derive from the non-deterministic nature of the algorithm, but all the clusters are changing their shape and position within the embedding.

On the images you can see results of alternating transforms on epoch e=83 / 82 for fixed l and i.

![testing_repeated_umap_0](https://user-images.githubusercontent.com/20558123/58800903-6347fa80-8609-11e9-8db6-a85129b5ca80.png)
![testing_repeated_umap_1](https://user-images.githubusercontent.com/20558123/58800900-6347fa80-8609-11e9-8a1c-6bb8fd3ec201.png)
![testing_repeated_umap_2](https://user-images.githubusercontent.com/20558123/58800897-6347fa80-8609-11e9-8648-ca5a5e620199.png)
![testing_repeated_umap_3](https://user-images.githubusercontent.com/20558123/58800896-62af6400-8609-11e9-8858-b2a1f54d98a3.png)
![testing_repeated_umap_4](https://user-images.githubusercontent.com/20558123/58800893-62af6400-8609-11e9-94e4-30436587481e.png)
![testing_repeated_umap_5](https://user-images.githubusercontent.com/20558123/58800894-62af6400-8609-11e9-8e91-3128146b3a9e.png)
![testing_repeated_umap_6](https://user-images.githubusercontent.com/20558123/58800904-6347fa80-8609-11e9-9710-ca8fbad682a5.png)

Do you know, why the clusters are shifting so much and how I could possibly prevent this?
Thank you very much for your time.

**Comment by lmcinnes** [comment]
It should not change things like this  -so there is a bug somewhere, possibly accidentally introduced when fixing some other issues. I'm pretty busy at the moment, but if you could ping me again in a week or two I will try to look at this properly.

**Comment by Kenneth-Schroeder** [comment]
Thanks for the fast response. There's no hurry and I will be happy to remind you :-)

**Comment by Kenneth-Schroeder** [comment]
Just a short update:
After a little more research I believe that the reason for this issue might be that UMAP is transductive, as you stated [here](https://github.com/lmcinnes/umap/issues/55#issuecomment-371170888).
However, this confuses me since in your [talk](https://youtu.be/nq6iPZVUxZU?t=1205) you say that new unseen points can be embedded using UMAP.

To clarify: In my experiment above I only fit the UMAP instance on the embedding of layer l of epoch 83. The embedding of layer l of epoch 82 might contain new (unseen) points for the UMAP reducer, which I just passed into the transform function.

**Comment by lmcinnes** [comment]
It should work as you are expecting, but for some reason something is
astray. It is essentially transductive, but there are some ways we can deal
with new data (even though we don't get an explicit map).

**Comment by Kenneth-Schroeder** [comment]
Here's a friendly reminder, just as you wished :-)

**Comment by lmcinnes** [comment]
Thanks for the reminder. I did some digging and it seems like the way this might occur is if the number of points you are transforming is exactly equal to the number of points in the initial embedding -- there is a check in the code to see whether you are trying to transform new points (and hence holds reference points fixed) but that only checks if the numbers of points are the same (because that was cheap and easy). Is it possible that this is the problem? If so I can try to make a still cheap but more robust check. Otherwise it might be time to dig deeper and figure out what is going on.

**Comment by Kenneth-Schroeder** [comment]
That sounds exactly right! I am using 5000 points for each class each time (35000 points in total) because the counts would be badly distributed otherwise. Nice catch :-)

**Comment by lmcinnes** [comment]
Ah, I see. I'll try to make the test a little more robust against this, ideally without causing too much overhead.

**Comment by ThanhHa12** [comment]
Hi lmcinnes,
not sure, if I understand this correctly but I think I have a similar issue. When using umap, every time I use a different random state, my clusters look very different. How to interpret this? Which metric should I use to measure the purity of the clusters?
